<a href="https://colab.research.google.com/github/SebastianPerezCalle/SENALES-Y-SISTEMAS-2025-1/blob/main/Python%20Basico/3_Loops.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Loops

In Python, we have two structures for loops: the for loop and the while loop (there is no do-while loop).


It is important to have a clear understanding of the concepts of a counter and an accumulator, as they are essential when working with any type of loop.

A counter is a variable that keeps track of the number of times a particular event or condition occurs. It is commonly used in loops to control the iteration process. The counter is usually initialized before the loop starts and is incremented or decremented with each iteration.

An accumulator, on the other hand, is a variable that accumulates or collects values over multiple iterations of a loop. It is often used to calculate a sum, a product, or any other aggregate value based on the values obtained during each iteration of the loop. The accumulator is typically initialized before the loop and updated within the loop body.

Both counters and accumulators play crucial roles in controlling and managing the flow of a loop, allowing you to perform specific actions based on the current state or progress of the loop iteration.

# While


They are also called indeterminate loops, and they are used when we know the condition that must be met to end the loop, but we do not know how many times or how long it will take.

In [1]:
condition = 1
while condition:
  print('Instructions inside the loop')
  condition = 0
print('Instructions outside the loop')

Instructions inside the loop
Instructions outside the loop


The structure of the loop is very simple, as the only thing that needs to be specified is the condition that must be met to continue in the loop. Essentially, it functions similarly to a conditional if() statement, except that the code inside the loop will be executed as many times as necessary as long as the condition is true.

Since the loop can be understood as a conditional statement, it works in the same way for boolean values, as well as for the values 1 and 0 explained in other materials.

In [2]:

a = 1
while a:
  print("Choose an option:\n1. ---\n2. ---\n0. Exit\n\n")
  a = int(input())


Choose an option:
1. ---
2. ---
0. Exit


1
Choose an option:
1. ---
2. ---
0. Exit


0



In the previous example, we relied on the value of 'a' itself as the condition for the loop.

However, assigning a value beforehand can sometimes cause issues.

Another way to do it is by directly performing a comparison as the loop condition.

In [3]:
number = int(input('Enter an integer number. 0 to finish: '))

while number != 0:
    number = int(input('Enter an integer number. 0 to finish: '))

print('End of the program.')

Enter an integer number. 0 to finish: 0
End of the program.


Then, it is possible to use infinite loops like while(True) or while(1) if something is required to run all the time.

Though it is not ideal to use it with the while() loop, it exists the statement break, which allows to get out of the loop in the previous scenaries, in case that it is critical not to run the remaining lines before ending the loop.

A common error using while() loops consists of not updating the control variable, thus creating a non intentional infinite loop (a bug in the code).

In [4]:
i = 0

# Run this loop as long as i is less than 15
while i < 15:
    # Show a message
    print("¡Hello world!")
    i+=1



¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!
¡Hello world!


In [5]:
4%2 #the % operator give you the residual between the numbers!

0

# For

for() loops are used to have complete control of how many iterations will be done.

It is necessary to have an "initial" value and a "final" value.

Optionally we can have a determined step size.

Opposite to C, for example, where we have all of these arguments in the loop statement, in Python we have to use the [range](https://www.w3schools.com/python/ref_func_range.asp) function.

In [6]:
#for counter_variable in range(initial_value, final_value, step_size):

    # code within the loop


# code out of the loop

For example, it we want a code to show all the even numbers from 900 to 1000, we have:

In [7]:
for i in range(900, 1000, 2):
    print(i)

900
902
904
906
908
910
912
914
916
918
920
922
924
926
928
930
932
934
936
938
940
942
944
946
948
950
952
954
956
958
960
962
964
966
968
970
972
974
976
978
980
982
984
986
988
990
992
994
996
998


We can see how it printed until 998 (as it was shown, we take until one number prior; if we would like to include 1000, it should be at least until 1001).


We can take advantage of the default arguments of the range() function for the following example, where we take the numbers from 0 to 10000 and we count all of them which are multiple of 33.

In [8]:
counter = 0

for i in range(10000):
    if (i % 33 == 0): # residual == 0?
        counter += 1

#show counter
print(counter)

304


For loops are used for excellence to go over arrays, lists or matrices, as their dimensions are fixed values, and allow to iterate its elements at please.

In [9]:
a = [8, 10, 23, 17, 15] #list of numbers

for i in range(len(a)):
  print(a[i])

8
10
23
17
15


It's important to note how len(a) returns the number of elements, but in Python, the index goes from zero, that's why all of the elements were printed. Let's see the difference:

In [10]:
len(a)

5

In [11]:
for i in range(4):
  print(a[i])

8
10
23
17


In [12]:
for i in range(5):
  print(a[i])

8
10
23
17
15


Then, as we have nested conditionals, we can also have nested loops, where each iteration of the external loops makes one full iteration of the internal loop. Combining it with the things said before, let's see an example with a matrix (list of lists):

In [13]:
M = [[1,3,4],[5,6,7],[8,9,2]]
M

[[1, 3, 4], [5, 6, 7], [8, 9, 2]]

In [14]:
for i in range(3): # Rows
  for j in range(3): # Columns
    print("Row " + str(i) + ", Column " + str(j) + ": " + str(M[i][j]))

Row 0, Column 0: 1
Row 0, Column 1: 3
Row 0, Column 2: 4
Row 1, Column 0: 5
Row 1, Column 1: 6
Row 1, Column 2: 7
Row 2, Column 0: 8
Row 2, Column 1: 9
Row 2, Column 2: 2


# Examples

Write a program that prints the numbers from 1 to 100, but for multiples of three, print "Fizz" instead of the number, and for multiples of five, print "Buzz".

For numbers that are multiples of both three and five, print "FizzBuzz"

In [15]:
for i in range(1, 101):
    if i % 3 == 0 and i % 5 == 0: #and operator included
        print("FizzBuzz")
    elif i % 3 == 0:
        print("Fizz")
    elif i % 5 == 0:
        print("Buzz")
    else:
        print(i)

1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
19
Buzz
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz
31
32
Fizz
34
Buzz
Fizz
37
38
Fizz
Buzz
41
Fizz
43
44
FizzBuzz
46
47
Fizz
49
Buzz
Fizz
52
53
Fizz
Buzz
56
Fizz
58
59
FizzBuzz
61
62
Fizz
64
Buzz
Fizz
67
68
Fizz
Buzz
71
Fizz
73
74
FizzBuzz
76
77
Fizz
79
Buzz
Fizz
82
83
Fizz
Buzz
86
Fizz
88
89
FizzBuzz
91
92
Fizz
94
Buzz
Fizz
97
98
Fizz
Buzz


Write a program that prompts the user to enter a positive integer, and then calculates the sum of all integers from 1 to the entered number.

In [16]:
num = int(input("Enter a positive integer: "))
sum = 0

# check if num is positive
if num <= 0:
    print("Invalid input! Please enter a positive integer.")
else:
    # calculate the sum of integers from 1 to num
    for i in range(1, num+1):
        sum += i

    print("The sum of integers from 1 to {} is: {}".format(num, sum))

Enter a positive integer: 0
Invalid input! Please enter a positive integer.


Generates a random number between 1 and 100, and prompts the user to guess the number.

If the guess is too high, the program should print "Too high", if the guess is too low, the program should print "Too low", and if the guess is correct, the program should print "Congratulations!" and exit the loop:

In [17]:
import random

# generate a random number between 1 and 100
num = random.randint(1, 100)

# loop until the user guesses the number
while True:
    guess = int(input("Guess the number (between 1 and 100): "))

    # check if the guess is correct
    if guess == num:
        print("Congratulations! You guessed the number.")
        break

    # check if the guess is too high
    elif guess > num:
        print("Too high! Guess again.")

    # check if the guess is too low
    else:
        print("Too low! Guess again.")

Guess the number (between 1 and 100): 0
Too low! Guess again.
Guess the number (between 1 and 100): 0
Too low! Guess again.
Guess the number (between 1 and 100): 5
Too low! Guess again.
Guess the number (between 1 and 100): 100
Too high! Guess again.
Guess the number (between 1 and 100): 1
Too low! Guess again.
Guess the number (between 1 and 100): 222
Too high! Guess again.
Guess the number (between 1 and 100): 50
Too low! Guess again.
Guess the number (between 1 and 100): 99
Too high! Guess again.
Guess the number (between 1 and 100): 70
Too low! Guess again.
Guess the number (between 1 and 100): 60
Too low! Guess again.
Guess the number (between 1 and 100): 55
Too low! Guess again.
Guess the number (between 1 and 100): 51
Too low! Guess again.
Guess the number (between 1 and 100): 80
Too low! Guess again.
Guess the number (between 1 and 100): 85
Too low! Guess again.
Guess the number (between 1 and 100): 86
Too low! Guess again.
Guess the number (between 1 and 100): 90
Congratulatio

Write a program that generates a multiplication table for all numbers between 1 and 10. (see [.format()](https://www.geeksforgeeks.org/python-string-format-method/) method)

In [18]:
for i in range(1, 11):
    for j in range(1, 11):
        print("{:4d}".format(i*j), end="")
    print()

   1   2   3   4   5   6   7   8   9  10
   2   4   6   8  10  12  14  16  18  20
   3   6   9  12  15  18  21  24  27  30
   4   8  12  16  20  24  28  32  36  40
   5  10  15  20  25  30  35  40  45  50
   6  12  18  24  30  36  42  48  54  60
   7  14  21  28  35  42  49  56  63  70
   8  16  24  32  40  48  56  64  72  80
   9  18  27  36  45  54  63  72  81  90
  10  20  30  40  50  60  70  80  90 100


## Exercise 1

Write a code that ask the user to enter an integer number every time, and the process will be repeated with more numbers until the user enters a zero.

At the end (when the zero is entered), show the amount of numbers entered.

## Exercise 2
Use the input() function to ask for the user to enter one integer number. Only if the number is even, it will be accumulated.

The program must come to an end if the sum of the numbers surpasses 50 or if the user enters a maximum of 10 numbers.

Developed by:

- Brian Patino Agudelo - bpatinoa@unal.edu.co

- Andrés Marino Álvarez Meza - amalvarezme@unal.edu.co

- César Germán Castellanos Domínguez - cgcastellanosd@unal.edu.co

Ejercicio 1: Contar cuántos números se ingresaron hasta que se introduce un cero

In [19]:
# Programa que cuenta cuántos números se ingresan hasta que se introduce un cero
contador = 0

while True:
    numero = int(input("Ingrese un número entero (0 para terminar): "))
    if numero == 0:
        break
    contador += 1

print("Cantidad de números ingresados:", contador)


Ingrese un número entero (0 para terminar): 50
Ingrese un número entero (0 para terminar): 0
Cantidad de números ingresados: 1


Ejercicio 2: Acumular números pares hasta superar 50 o alcanzar 10 intentos

In [22]:
# Programa que acumula números pares hasta que la suma supere 50 o se ingresen 10 números
suma = 0
contador = 0

while suma <= 50 and contador < 10:
    numero = int(input("Ingrese un número entero: "))
    contador += 1
    if numero % 2 == 0:
        suma += numero

print("Programa finalizado.")
print("Suma total de números pares acumulados:", suma)
print("Cantidad total de números ingresados:", contador)


Ingrese un número entero: 15
Ingrese un número entero: 28
Ingrese un número entero: 59
Ingrese un número entero: 652
Programa finalizado.
Suma total de números pares acumulados: 680
Cantidad total de números ingresados: 4
